refactor: add `FeedbackDatasetBase` and `RemoteFeedbackDataset` while keeping `FeedbackDataset` just for local

alvarobartt opened the pull request:
# Description

This PR is still in progress, and the main idea is that we move the common functionality to be used on both local and remote (Argilla) datasets to `FeedbackDatasetBase`, while we keep `FeedbackDataset` for working locally with these datasets, and add `_ArgillaFeedbackDataset` to be instantiated internally via `FeedbackDataset.from_argilla` and as part of the return statement of `FeedbackDataset.push_to_argilla`, so as to split the behaviour on some operations such as `add_records`, since locally means adding those to a local list, while remotely that means pushing those to Argilla, to avoid having to call `push_to_argilla` right after every record addition.

Some more things are tackled as part of this refactoring and will be listed down below as soon as the PR is out of draft!

Closes #3456

**Type of change**

- [X] Refactor (change restructuring the codebase without changing functionality)
- [X] Improvement (change adding some improvement to an existing functionality)

**How Has This Been Tested**

- [x] Add unit tests for the different classes and methods, while ensuring backwards compatibility on everything related to the `FeedbackDataset`

**Checklist**

- [ ] I added relevant documentation
- [x] follows the style guidelines of this project
- [x] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK) (see text above)
- [x] I have added relevant notes to the CHANGELOG.md file (See https://keepachangelog.com/)

alvarobartt committed:
refactor: split `FeedbackDatasetBase` from `FeedbackDataset`

alvarobartt committed:
style: type-hints and imports

alvarobartt committed:
refactor: `ArgillaDatasetMixin` to `ArgillaToFromMixin`

alvarobartt committed:
refactor: `guidelines` is protected not private

alvarobartt committed:
refactor: remove `fetch_records` from `ArgillaToFromMixin`

alvarobartt committed:
fix: assign `id` but keep original schema

alvarobartt committed:
feat: remap class in `push_to_argilla`

Done to ensure consistency with the current approach, even though in upcoming releases no remapping will be done and the user will need to capture `push_to_argilla` method's output

alvarobartt committed:
fix: ensure `DeprecationWarning`s are shown

alvarobartt committed:
style: add `FeedbackRecord` under `TYPE_CHECKING`

alvarobartt committed:
refactor(wip): add `_ArgillaFeedbackDataset`

alvarobartt committed:
refactor: rename `ArgillaFeedbackDataset` to `_ArgillaFeedbackDataset`

alvarobartt committed:
Merge branch 'develop' of github.com:argilla-io/argilla into refactor/split-local-and-remote-dataset

alvarobartt committed:
fix: add missing `client` arg in `from_argilla`

alvarobartt committed:
feat: add `get_metrics` function in SDK (#3469)

# Description

This PR adds a function in the SDK named `get_metrics` that calls `GET
/api/v1/me/datasets/{dataset_id}/metrics` to get the metrics of a
certain dataset from a certain user, meaning that each user would just
see what can be seen according to its permissions.

This PR is created on top of #3465 since we will need to get the
`records.count` to override the magic method `__len__` in
`_ArgillaFeedbackDataset` so that the length of a dataset is directly
retrieved from the record count of the dataset, since we're no longer
keeping local data when working with a remote dataset.

**Type of change**

- [X] New feature (non-breaking change which adds functionality)

**How Has This Been Tested**

- [X] Add unit tests for the `get_metrics` function

**Checklist**

- [ ] I added relevant documentation
- [X] follows the style guidelines of this project
- [X] I did a self-review of my code
- [ ] I made corresponding changes to the documentation
- [X] My changes generate no new warnings
- [X] I have added tests that prove my fix is effective or that my
feature works
- [ ] I filled out [the contributor form](https://tally.so/r/n9XrxK)
(see text above)
- [x] I have added relevant notes to the CHANGELOG.md file (See
https://keepachangelog.com/)

---------

Co-authored-by: Francisco Aranda <francis@argilla.io>

alvarobartt committed:
feat: add `__len__` and `__getitem__`

alvarobartt committed:
Merge branch 'develop' into refactor/split-local-and-remote-dataset

alvarobartt committed:
refactor: remove `_updated` field from `FeedbackRecord`

alvarobartt committed:
feat: add `_ArgillaFeedbackRecord`

alvarobartt committed:
fix: `set_suggestion` type-hint

alvarobartt committed:
feat: add `_ArgillaFeedbackRecords`

alvarobartt committed:
fix: add `arbitrary_types_allowed` for `client`

alvarobartt committed:
feat: use `id` and `questions` in `_ArgillaFeedbackRecords`

frascuchon commented on src/argilla/client/feedback/dataset/local.py:
```diff
@@ -0,0 +1,103 @@
+        warnings.warn(
+            "As the current `FeedbackDataset` is stored locally and is pushed to Argilla,"
+            " the method `fetch_records` won't do anything. If you want to fetch the records"
+            " from Argilla, make sure you're using an `_ArgillaFeedbackDataset`, either by"
```
I think we can define this class as public `ArgillaFeedbackDataset`

gabrielmbmb replied:
I agree, it will be used directly by the user.

alvarobartt replied:
Hm but we don't want users to actually use it, we want to use it just internally, right?

alvarobartt replied:
It has been renamed to `RemoteFeedbackDataset` as internally suggested by @frascuchon

frascuchon commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+        show_progress: bool = False,
+    ) -> _ArgillaFeedbackDataset:
+        """Pushes the `FeedbackDataset` to Argilla. If the dataset has been previously pushed to Argilla, it will be updated
+        with the new records.
```
The second part of the message is outdated, right?
```suggestion
        """Pushes the `FeedbackDataset` to Argilla. 
```

frascuchon commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+
+        Args:
+            name: the name of the dataset to push to Argilla. If not provided, the `argilla_id` will be used if the dataset
+                has been previously pushed to Argilla.
```
```suggestion
            name: the name of the dataset to push to Argilla. 
```

frascuchon commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+            fields=fields,
+            questions=questions,
+            guidelines=self.guidelines,
+        )
```
This is for backward compatibility or is there any other reason?

Ideally, this could be removed in future releases, to allow publish the same dataset with different names/workspaces:
```python
for ws in ...:
   ds.push_to_argilla(name, ws)
```

alvarobartt replied:
Yes, it's for backward compatibility, as otherwise, users would face a breaking change and would need to either capture the output or call `FeedbackDataset.from_argilla` right after the push

frascuchon commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+        *,
+        workspace: Optional[str] = None,
+        id: Optional[str] = None,
+        with_records: bool = True,  # TODO(alvarobartt): deprecate `with_records`
```
If we set to `None` as default, we can detect that a value is provided and show the warning properly.
```suggestion
        with_records: Optional[bool] = None,  # TODO(alvarobartt): deprecate `with_records`
```

alvarobartt replied:
Yes, docstrings and messages have not been revisited yet, just the functionality

frascuchon commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+            >>> rg.init(...)
+            >>> dataset = rg.FeedbackDataset.from_argilla(name="my_dataset")
+        """
+        if with_records:
```
```suggestion
        if with_records is not None:
```

frascuchon commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,220 @@
+warnings.simplefilter("always", DeprecationWarning)
+
+
+class _ArgillaFeedbackRecords:
```
I think we can define as a public class
```suggestion
class ArgillaFeedbackRecords:
```

gabrielmbmb replied:
I think it is fine as it is. The user will not use this class directly, but through the `ArgillaFeedbackDataset`, right?

alvarobartt replied:
It has been renamed to `RemoteFeedbackDataset` as internally suggested by @frascuchon

gabrielmbmb commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -298,19 +211,44 @@ def question_by_name(self, name: str) -> AllowedQuestionTypes:
+                else:
+                    raise ValueError(
+                        f"Expected `records` to be a list of `dict` or `FeedbackRecord`, got type {type(record)} instead."
+                    )
```
We could move this block to a function.

gabrielmbmb commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -298,19 +211,44 @@ def question_by_name(self, name: str) -> AllowedQuestionTypes:
+            records = [records]
+        else:
+            raise ValueError(
+                f"Expected `records` to be a `dict` or `FeedbackRecord`, got type {type(records)} instead."
```
```suggestion
                f"Expected `records` to be a `dict` or `FeedbackRecord`, got type `{type(records)}` instead."
```

gabrielmbmb commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -298,19 +211,44 @@ def question_by_name(self, name: str) -> AllowedQuestionTypes:
+            except ValidationError as e:
+                raise ValueError(
+                    f"`FeedbackRecord.fields` does not match the expected schema, with exception: {e}"
+                ) from e
```
We could move this block to a function.

gabrielmbmb commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+                " or just call this function and then `from_argilla` to retrieve the"
+                " `FeedbackDataset` from Argilla.",
+                DeprecationWarning,
+            )
```
when are we going to deprecate the current logic? `1.15.0`? once we know we should add it to this warning message

alvarobartt replied:
I guess this was not discussed that's why I didn't specify a version, but yes, ideally we should specify the version 👍🏻

gabrielmbmb commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+                raise Exception(
+                    f"Failed while deleting the `FeedbackDataset` with ID '{dataset_id}' from Argilla with"
+                    f" exception: {e}"
+                ) from e
```
why we don't have this as a method already?

alvarobartt replied:
It's on the works, PR coming later today 👍🏻

gabrielmbmb commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+                raise Exception(
+                    f"Failed while adding the field '{field.name}' to the `FeedbackDataset` in Argilla with"
+                    f" exception: {e}"
+                ) from e
```
We could move this block to a function.

gabrielmbmb commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+                raise Exception(
+                    f"Failed while adding the question '{question.name}' to the `FeedbackDataset` in Argilla"
+                    f" with exception: {e}"
+                ) from e
```
we could move this block to a function.

gabrielmbmb commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+                delete_dataset(dataset_id=argilla_id)
+                raise Exception(
+                    f"Failed while adding the records to the `FeedbackDataset` in Argilla with exception: {e}"
+                ) from e
```
we could move this block to a function.

gabrielmbmb commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+                    f"Field '{field.name}' is not a supported field in the current Python package version,"
+                    " supported field types are: `TextField`."
+                )
+            fields.append(field)
```
This block could be a function.

gabrielmbmb commented on src/argilla/client/feedback/dataset/mixins.py:
```diff
@@ -0,0 +1,296 @@
+                    f"Question '{question.name}' is not a supported question in the current Python package"
+                    f" version, supported question types are: `{'`, `'.join([arg.__name__ for arg in AllowedQuestionTypes.__args__])}`."
+                )
+            questions.append(question)
```
This block could be a function.

alvarobartt committed:
feat: add `update` & `DeprecationWarning` for `set_suggestions`

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -371,14 +310,16 @@ def unify_responses(
+    # TODO(alvarobartt,davidberenstein1957): we should consider having something like
+    # `export(..., training=True)` to export the dataset records in any format, replacing
+    # both `format_as` and `prepare_for_training`
     def prepare_for_training(
```
Does make sense to have this functionality linked to local datasets? If the normal workflow is  1. download what you want from argilla locally and 2. train your model, maybe makes sense to prepare the training data only from a local version of the feedback dataset.

alvarobartt replied:
I've ensured that it's backwards compatible with the current approach, so that the records are automatically pulled from Argilla if we want to use a `FeedbackDataset` in Argilla i.e. `RemoteFeedbackDataset`

frascuchon replied:
👍  In any case, I would say to warn users if the `prepare_for_training` method is done using the `RemoteArgillaDataset`. Since this will change in the future, and all the training stuff will be applied using local copies of datasets.

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -336,6 +274,7 @@ def format_as(self, format: Literal["datasets"]) -> "Dataset":
             return self._huggingface_format(self)
         raise ValueError(f"Unsupported format '{format}'.")
 
+    # TODO(davidberenstein1957): detatch unification into a mixin
```
Same here, maybe this makes sense only for local copies of the datasets.

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -66,63 +65,8 @@ class FeedbackDataset(ArgillaDatasetMixin, HuggingFaceDatasetMixin):
     """
 
-    argilla_id: Optional[UUID] = None
-
```
If a user is using the property, this change may break the code. Maybe just a property with 2 different implementations?

```python
@property
def argilla_id() -> UUID:
   raise NotImplementedError()

# for remote dataset:
def argilla_id() -> UUID:
   warnings.warning("use self.id instead...")
   return self.id

# for local dataset
def argilla_id() -> UUID:
   warnings.warning("local datases don't track remote argilla info...")
   return None
```

alvarobartt replied:
Done, thanks!

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -298,19 +211,44 @@ def question_by_name(self, name: str) -> AllowedQuestionTypes:
 
-    def iter(self, batch_size: Optional[int] = FETCHING_BATCH_SIZE) -> Iterator[List[FeedbackRecord]]:
-        """Returns an iterator over the records in the dataset.
+        if self._fields_schema is None:
+            self._fields_schema = generate_pydantic_schema(self.fields)
```
Does it mean that once the pydantic schema is generated the field_schema is not changed after a field setting update? Maybe we can just generate each time the method is called:
```suggestion
        fields_schema  = generate_pydantic_schema(self.fields)
```
Ow, are we using the self._fields_schema outside this method?

alvarobartt replied:
No we're not, but ideally we want to re-use it, instead of calculating it every time, and since we don't support field settings updates for the moment I'd say we can leave it as is

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -298,19 +211,44 @@ def question_by_name(self, name: str) -> AllowedQuestionTypes:
-            yield self.records[i : i + batch_size]
+        for record in records:
+            try:
+                self._fields_schema.parse_obj(record.fields)
```
```suggestion
                fields_schema.parse_obj(record.fields)
```

alvarobartt committed:
fix: add `argilla_id` for backwards compatibility

alvarobartt committed:
fix: deprecate `fetch_records` in `prepare_for_training`

alvarobartt committed:
fix: update `warnings`  in `set_suggestions` and `fetch_records`

alvarobartt committed:
feat: add `pull` to restore `FeedbackDataset` from Argilla

alvarobartt committed:
fix: add `DeprecationWarning` in `fetch_records`

alvarobartt committed:
test: update `FeedbackDataset` unit tests

alvarobartt committed:
refactor: rename `_ArgillaFeedbackDataset` to `RemoteFeedbackDataset`

alvarobartt committed:
fix: default `with_records` to `None`

Co-authored-by: Francis Aranda <francis@argilla.io>

alvarobartt committed:
fix: remove wrong `" f"`

alvarobartt committed:
fix(docs): update `ArgillaToFromMixin` docstrings

alvarobartt committed:
Merge branch 'develop' of github.com:argilla-io/argilla into refactor/split-local-and-remote-dataset

frascuchon commented on CHANGELOG.md:
```diff
@@ -18,6 +18,7 @@ These are the section headers that we use:
 
 ### Added
 
+- Added `get_metrics` function in the SDK to get `FeedbackDataset` metrics calling `GET /api/v1/me/{dataset_id}/metrics` endpoint ([#3469](https://github.com/argilla-io/argilla/pull/3469))
```
I think we can remove this line since is too techie detail for a changelog 
```suggestion
```

By the way, we can add other in the changed and deprecated sections, referring the new `add_records` behaviour. or the push_to_argilla and other deprecated methods

alvarobartt replied:
Thanks, already done!

alvarobartt committed:
docs: update `CHANGELOG.md`

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,247 @@
+            " fetched, and never stored locally. You can either loop over `records`"
+            " or access them by index, and those will be fetched from Argilla on the"
+            " fly."
+        )
```
I think it's a bit weird that if you print `ArgillaFeedbackDataset.records` you get this, it's more like a warning message. This method could return something like:

`[RemoteFeedbackRecord(...), RemoteFeedbackRecord(...), ...] lazily fetched from 'https://dev.argilla.io'`

frascuchon replied:
Yes, I agree with @gabrielmbmb. Maybe we can just print some of the records:
```suggestion
        return f"{self.records[:3]!r}..."
```

frascuchon replied:
Even if I'm not much comfortable making remote calls in `__repr__` method. But can be useful to users.

alvarobartt replied:
I think this is not optimal and can also be confusing 😩 Maybe we can just go with @gabrielmbmb's suggestion to show something like `[RemoteFeedbackRecord(...), RemoteFeedbackRecord(...), ...] lazily fetched from 'https://dev.argilla.io'`?

alvarobartt replied:
Or maybe just return `None` and a `warnings.warn` instead

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,247 @@
+                yield self.__parse_record(record)
+            current_batch += 1
+
+            if len(batch.items) < FETCHING_BATCH_SIZE:
+                break
```
we can move this before the for loop

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,247 @@
+            "`argilla_id` is deprected in favor of `id` and will be removed in a future"
+            " release. Please use `id` instead.",
+            DeprecationWarning,
+        )
```
add version in which it will be deprecated once we know

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,247 @@
+        return self._workspace
+
+    def __repr__(self) -> str:
+        return f"<ArgillaFeedbackDataset id={self.id} name={self.name} workspace={self.workspace}>"
```
maybe it would be good to add the URL of the Argilla server

frascuchon replied:
Yes, I would say that having `dataset.url` property  could be great

gabrielmbmb replied:
Also, it should be '<RemoteFeedbackDataset ...>`, right? to be aligned with the class name

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,247 @@
+        return self.records.__len__()
+
+    def __iter__(self) -> Iterator[RemoteFeedbackRecord]:
+        return self.records.__iter__()
```
```suggestion
        yield from self.records
```

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,247 @@
+            " as the additions, deletions and/or updates over a `FeedbackDataset` in Argilla"
+            " are being tracked automatically, so there's no need to explicitly push them.",
+            DeprecationWarning,
+        )
```
add version in which this will be deprecated once we know

gabrielmbmb commented on src/argilla/client/feedback/schemas/records.py:
```diff
@@ -204,13 +206,48 @@ def set_suggestions(
+            "`set_suggestions` is deprected in favor of `update` and will be removed in a future"
+            " release. Please use `update` instead.",
+            DeprecationWarning,
+        )
```
add version in which this will get deprecated once we know

gabrielmbmb commented on src/argilla/client/feedback/schemas/records.py:
```diff
@@ -204,13 +206,48 @@ def set_suggestions(
+            "`set_suggestions` is deprected in favor of `update` and will be removed in a future"
+            " release. Please use `update` instead.",
+            DeprecationWarning,
+        )
```
add version in which this will get deprecated once we know

gabrielmbmb reviewed (approved):
Hey @alvarobartt, LGTM :) I have left you some comments. Also, if you could add the `stacklevel=1` to the `warnings.warn` calls, it would be good.

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -230,45 +175,14 @@ def __init__(
-        if isinstance(key, int) and len(self.records) < key:
-            raise IndexError(f"This dataset contains {len(self)} records, so index {key} is out of range.")
-        return self.records[key]
+        self.records: List[FeedbackRecord] = []
```
Should not we move this to the `FeedbackDataset` class? The `ArgillaFeedbackDataset` does not need a local list of records right?

Maybe the base class can just define an abstract property `records`, letting each implementation resolver in the proper way

```suggestion
   @property
   def records(self) -> Iterable[FeedbackRecord]:
      raise NotImplementedError()
```

alvarobartt replied:
It was defined that way because we're defining the `unify_responses` and `prepare_for_training` methods in the `FeedbackDatasetBase`, but I guess we can just define a property, but note that the property itself it's not an iterable when working with local datasets, as it's just a list, but I'll figure that out 👍🏻

frascuchon commented on src/argilla/client/feedback/dataset/local.py:
```diff
@@ -0,0 +1,113 @@
+            " value from `FeedbackDataset.push_to_argilla`.",
+        )
+
+    def add_records(
```
A question here. For local datasets, we could do something like:
```python
for data in ...:
   dataset.records.append(FeedbackRecord(**data))
``` 
to add records, right? I mean, it's possible right now, right? Maybe we should define the records as a property based on a protected attribute `__records` 


```python
@property
def records(self) -> Iterable[FeedbackRecord]:
   return (r for r in self.__records)
```

alvarobartt replied:
Sure, I can make `records` a property

alvarobartt committed:
Apply suggestions from code review

Co-authored-by: Gabriel Martin <gabriel@argilla.io>

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -298,19 +212,44 @@ def question_by_name(self, name: str) -> AllowedQuestionTypes:
-    def records(self) -> List[FeedbackRecord]:
-        """Returns the all the records in the dataset."""
-        return self._records + self._new_records
+    def _validate_records(
```
Not for this PR but, I think it make sense to validate records also based on dataset questions. Maybe records contain incompatible responses for configured dataset questions

alvarobartt replied:
Ok makes sense to add validations for both responses and suggestions, I'll do so in an upcoming PR 😄

frascuchon commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,249 @@
+            " release. Please use `id` instead.",
+            DeprecationWarning,
+        )
+        return self.id
```
```suggestion
        return str(self.id)
```

alvarobartt replied:
Why not return the `UUID`?

alvarobartt committed:
Apply suggestions from code review

Co-authored-by: Francis Aranda <francis@argilla.io>

alvarobartt committed:
feat: add `url` in `RemoteFeedbackDataset`

Co-authored-by: Gabriel Martin <gabriel@argilla.io>

alvarobartt committed:
Merge branch 'develop' into refactor/split-local-and-remote-dataset

alvarobartt committed:
Merge branch 'refactor/split-local-and-remote-dataset' of github.com:argilla-io/argilla into refactor/split-local-and-remote-dataset

alvarobartt committed:
fix: `argilla_id` return type-hint

Co-authored-by: Francis Aranda <francis@argilla.io>

alvarobartt committed:
fix: return type-hints in double-quotes when needed

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,257 @@
+        return f"{self.client.base_url}/dataset/{self.id}"
+
+    def __repr__(self) -> str:
+        return f"<ArgillaFeedbackDataset id={self.id} name={self.name} workspace={self.workspace}>"
```
```suggestion
        return f"<FeedbackDataset id={self.id} name={self.name} workspace={self.workspace} url={self.url}>"
```

frascuchon commented on CHANGELOG.md:
```diff
@@ -40,6 +40,13 @@ These are the section headers that we use:
 - Updated `SearchEngine.search` method to allow searching records matching one of the response statuses provided ([#3359](https://github.com/argilla-io/argilla/pull/3359)).
+- After calling `FeedbackDataset.push_to_argilla`, the methods `FeedbackDataset.add_records` and `FeedbackRecord.set_suggestions` will automatically call Argilla with no need of calling `push_to_argilla` explicitly ([#3465](https://github.com/argilla-io/argilla/pull/3465)).
+
+### Deprecated
```
does make sense also add the `pull` method to fetch a local copy of a remote Argilla dataset?

alvarobartt replied:
Yes, adding it now!

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -230,45 +176,18 @@ def __init__(
-        return self.records[key]
+    @property
+    @abstractproperty
+    def records(self) -> Any:
```
Could be expected an `Iterable[FeedbackRecord]` here?

alvarobartt replied:
No because `records` is not an iterable, it's a `List[FeedbackRecord]` in `FeedbackDataset` while that's an instance of `RemoteFeedbackRecords` in `RemoteFeedbackDataset`

frascuchon commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,257 @@
+class RemoteFeedbackRecords:
+    def __init__(self, client: "httpx.Client", id: "UUID", questions: List["AllowedQuestionTypes"]) -> None:
+        self.client = client
+        self.id = id
```
If the id refers to the dataset id, I would say to explicitly call `dataset_id`. Otherwise can be hard to know which id is referring to.

```suggestion
    def __init__(self, client: "httpx.Client", dataset_id: "UUID", questions: List["AllowedQuestionTypes"]) -> None:
        self.client = client
        self.dataset_id = dataset_id
```

alvarobartt replied:
I agree, renamed that and made it private already!

frascuchon commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,257 @@
+class RemoteFeedbackRecords:
+    def __init__(self, client: "httpx.Client", id: "UUID", questions: List["AllowedQuestionTypes"]) -> None:
+        self.client = client
+        self.id = id
```
Do we need the `self.id` to be public?

alvarobartt replied:
Nope, making both `client` and `id` private now!

frascuchon commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,257 @@
+                record = fetched_records.items[0]
+                records.append(self.__parse_record(record))
+            else:
+                records.extend([self.__parse_record(record) for record in fetched_records.items])
```
Why we cannot do something like:
```suggestion
            records.extend([self.__parse_record(record) for record in fetched_records.items])
```

alvarobartt replied:
Thanks makes sense!

alvarobartt committed:
docs: add missing docstrings

frascuchon commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,257 @@
+        if isinstance(key, slice):
+            start, stop, step = key.indices(len(self))
+            if step is not None and step != 1:
+                return [self[i] for i in range(start, stop, step)]
```
This case will call several times the API to fetch records one-by-one right? I would say start by not supporting slices with `step` != 1 for remote datasets.

alvarobartt replied:
But what if that's what the user wants? At the end we want to expose the records so that the user can use it as a normal list but inside it's fetching from Argilla, so I think we can leave it as is, at the end it depends on the user to avoid doing that

frascuchon commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,257 @@
+            if key < 0:
+                key += len(self)
+            if key < 0 or key >= len(self):
+                raise IndexError("Index out of range")
```
Just to avoid several calls to compute the lenght
```suggestion
        offsets = []
        limit = None
        records_lenght= len(self)
        if isinstance(key, slice):
            start, stop, step = key.indices()
            if step is not None and step != 1:
                return [self[i] for i in range(start, stop, step)]
            if start < 0:
                start += records_lenght
            if stop < 0:
                stop += records_lenght
            if start < 0 or stop < 0:
                raise IndexError("Index out of range")
            limit = stop - start
            offsets = [start] if limit < FETCHING_BATCH_SIZE else list(range(start, stop, FETCHING_BATCH_SIZE))
        elif isinstance(key, int):
            if key < 0:
                key += records_lenght
            if key < 0 or key >= records_lenght:
                raise IndexError("Index out of range")
```

alvarobartt replied:
Thanks makes sense!

frascuchon commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,257 @@
+        return f"<ArgillaFeedbackDataset id={self.id} name={self.name} workspace={self.workspace}>"
+
+    def __len__(self) -> int:
+        return self._records.__len__()
```
Is this the same as `len(self._records)`?

alvarobartt replied:
Yes, but calling directly the magic method instead

alvarobartt committed:
Apply suggestions from code review

Co-authored-by: Francis Aranda <francis@argilla.io>

alvarobartt committed:
Apply suggestions from code review

Co-authored-by: Gabriel Martin <gabriel@argilla.io>

frascuchon commented on src/argilla/client/feedback/dataset/base.py:
```diff
@@ -298,19 +178,64 @@ def question_by_name(self, name: str) -> AllowedQuestionTypes:
+            raise ValueError(
+                "Expected `records` to be a `dict` or `FeedbackRecord`, got type" f" `{type(records)}` instead."
+            )
+        return records
```
I think we can simplify this method, and avoid parsing logic duplication.
```suggestion
        if isinstance(records, (dict, FeedbackRecord)):
            records = [records]

        if len(records) == 0:
            raise ValueError("Expected `records` to be a non-empty list of `dict` or `FeedbackRecord`.")
            
        new_records = []
        for record in records:
            if isinstance(record, dict):
                new_records.append(FeedbackRecord(**record))
            elif isinstance(record, FeedbackRecord):
                new_records.append(record)
            else:
                raise ValueError(
                    "Expected `record` to be a list of `dict` or `FeedbackRecord`,"
                    f" got type `{type(record)}` instead."
                )
        return new_records
```

alvarobartt committed:
Apply suggestions from code review

Co-author-by: Gabriel Martin <gabriel@argilla.io>

alvarobartt committed:
fix: `IndexError` message in `__getitem__`

alvarobartt committed:
fix: `RemoteFeedbackRecords.__getitem__` using wrong `limit`

alvarobartt committed:
Apply suggestions from code review

Co-authored-by: Francisco Aranda <francis@argilla.io>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

alvarobartt committed:
test(unit): add `FeedbackDatasetBase` unit tests

alvarobartt committed:
fix: `ValueError` message in `_parse_records`

gabrielmbmb commented on src/argilla/client/feedback/dataset/remote.py:
```diff
@@ -0,0 +1,368 @@
+            " fetched, and never stored locally. You can either loop over `records`"
+            " or access them by index, and those will be fetched from Argilla on the"
+            " fly.",
+        )
```
```suggestion
        warnings.warn(
            "The `records` of a `FeedbackDataset` in Argilla are being lazily"
            " fetched, and never stored locally. You can either loop over `records`"
            " or access them by index, and those will be fetched from Argilla on the"
            " fly.",
            stacklevel=1
        )
```

alvarobartt committed:
Apply suggestions from code review

Co-authored-by: Gabriel Martín Blázquez <gmartinbdev@gmail.com>

pre-commit-ci[bot] committed:
[pre-commit.ci] auto fixes from pre-commit.com hooks
for more information, see https://pre-commit.ci

alvarobartt committed:
fix(test): add missing fixtures & instantiate `FeedbackDatasetBase`

alvarobartt committed:
fix(test): partial strip of `Exception` match

Due to the same message being formatted differently depending on the Python version e.g. `abstract methods records` in Python 3.10, but `abstract method records` in Python 3.7

alvarobartt committed:
fix(test): remove `add_records(None)` test
